### PR TITLE
Add syscall sendmsg and recvmsg

### PR DESCRIPTION
**Pull request by StevenJiang1110**
This PR adds syscall sendmsg and recvmsg and revises syscall using `IoVec`.

### CODE REVIEW

**Comment by tatetian** [comment]
@lrh2000 Could you take a look at this PR? Thanks.

**Comment by lrh2000** [comment]
@tatetian 

> Could you take a look at this PR? Thanks.

I'll take a look, but I'm in CA, USA attending NSDI'24 this week, so the response may be delayed.

**Comment by tatetian** [comment]
> @tatetian
> 
> > Could you take a look at this PR? Thanks.
> 
> I'll take a look, but I'm in CA, USA attending NSDI'24 this week, so the response may be delayed.

Thanks for letting us know. Have a nice trip!

File: kernel/aster-nix/src/util/iovec.rs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+impl Iterator for IoVecIter {
+    type Item = Result<IoVec>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        while self.current < self.iovec_count {
+            let addr = self.start_addr + self.current * mem::size_of::<IoVec>();
+            self.current += 1;
+
+            let res = read_val_from_user::<IoVec>(addr);
```
Comment (lrh2000) [inline kernel/aster-nix/src/util/iovec.rs:79]: The current implementation of `IoVecIter` is wrong, as far as I can tell from [the Linux implementation](https://elixir.bootlin.com/linux/v6.8.7/source/lib/iov_iter.c#L1145). Instead of trying to copy them one by one when it is used, we need to copy all the `IoVec`s into kernel space at the beginning.
 - The latter approach actually causes problems. What error code should we return to the user application when `EFAULT` happens? The answer is inconsistent if the `EFAULT` happens on the first `IoVec` or in the middle.

File: framework/aster-frame/src/vm/memory_set.rs [hunk 2]
```diff
@@ -8,8 +8,8 @@ use crate::{
     arch::mm::{PageTableEntry, PageTableFlags},
     prelude::*,
     vm::{
-        is_page_aligned, VmAllocOptions, VmFrame, VmFrameVec, VmReader, VmWriter,
-        PHYS_MEM_BASE_VADDR, PAGE_SIZE,
+        is_page_aligned, VmAllocOptions, VmFrame, VmFrameVec, VmReader, VmWriter, PAGE_SIZE,
+        PHYS_MEM_BASE_VADDR,
```
Comment (lrh2000) [inline framework/aster-frame/src/vm/memory_set.rs:12]: Could you please remove the unrelated changes from this PR? I think rebasing the PR and resolving the conflicts will solve this directly.

File: kernel/aster-nix/src/util/iovec.rs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+    pub const fn len(&self) -> usize {
+        self.len
+    }
+
+    pub fn read_from_user(&self, buffer: &mut [u8]) -> Result<()> {
+        debug_assert!(buffer.len() == self.len);
+        if buffer.len() != self.len {
+            return_errno_with_message!(
+                Errno::EINVAL,
```
Comment (lrh2000) [inline kernel/aster-nix/src/util/iovec.rs:31]: There is a `debug_assert`, but I still want to ask, do we really want to expose kernel implementation bugs as (random) user-facing error codes?

Apart from that, how do you feel about revising the APIs as following:
```rust
pub fn read_from_user<R>(&self, consumer: FnOnce(&[u8]) -> Result<R>) -> Result <R> {
    /* ... */
}

pub fn write_from_user<R>(&self, producer: FnOnce(&mut [u8]) -> Result<R>) -> Result<R> {
    /* ... */
}
```
where the `consumer` consumes the data (e.g., by calling `try_sendto`) and the `producer` produces the data (e.g., by calling `try_recvfrom`).

File: kernel/aster-nix/src/net/socket/ip/datagram/mod.rs [hunk 4]
```diff
@@ -284,6 +284,68 @@ impl Socket for DatagramSocket {
...
+
+            // TODO: Block if the send buffer is full
+            let sent_bytes = self.try_sendto(&buffer, remote_endpoint, flags)?;
+            total_bytes += sent_bytes;
+        }
+
+        Ok(total_bytes)
+    }
+
+    fn recvmsg(&self, msg_hdr: &mut MessageHeader, flags: SendRecvFlags) -> Result<usize> {
```
Comment (lrh2000) [inline kernel/aster-nix/src/net/socket/ip/datagram/mod.rs:326]: I don't think it's clear which fields in `msg_hdr: &mut MessageHeader` are input and which are output.

Maybe we can have a better API for this function? Is it possible to put output fields into the return type?

File: kernel/aster-nix/src/util/net/socket.rs [hunk 5]
```diff
@@ -68,3 +69,33 @@ bitflags! {
...
+
+impl CUserMsgHdr {
+    pub fn read_socket_addr(&self) -> Result<Option<SocketAddr>> {
+        if self.msg_name == 0 {
+            return Ok(None);
+        }
+
+        let socket_addr = read_socket_addr_from_user(self.msg_name, self.msg_namelen as usize)?;
+        Ok(Some(socket_addr))
+    }
```
Comment (lrh2000) [inline kernel/aster-nix/src/util/net/socket.rs:100]: Perhaps we have a method that creates the `IoVecIter` here? We invoke the long expression `IoVecIter::new(c_user_msghdr.msg_iov, c_user_msghdr.msg_iovlen as usize)` multiple times.

File: kernel/aster-nix/src/net/socket/ip/datagram/mod.rs (hunk 4 above)
Comment (StevenJiang1110) [reply kernel/aster-nix/src/net/socket/ip/datagram/mod.rs]: This API has been refactored to return `(usize, MessageHeader)`

File: kernel/aster-nix/src/util/iovec.rs (hunk 3 above)
Comment (StevenJiang1110) [reply kernel/aster-nix/src/util/iovec.rs]: > do we really want to expose kernel implementation bugs as (random) user-facing error codes?

I think we should not expose these errors to users. I have refactored the code, use `assert_eq` instead of `debug_assert`, so we don't need to return error code.

As for your suggesting API changes, it's mostly reasonable to me. But I have one concern:

These APIs suggest the kernel buffer should be allocated by `IoVec` itself, and then used by `producer` or `consumer`. but indeed, `IoVec` may not have enough information to decide which type of memory to allocate, for example, allocate heap memory, or `VmFrames`, or DMA-capable memory. Only the user knows what the exact memory type they do need.

File: kernel/aster-nix/src/util/iovec.rs (hunk 1 above)
Comment (StevenJiang1110) [reply kernel/aster-nix/src/util/iovec.rs]: I have refactored the implementation to read all `IoVec`s at beginning. But I didn't find any error number other than `EFAULT`. (The only exception is that we return `EINVAL` if iovec.len < 0).

Comment (lrh2000) [reply kernel/aster-nix/src/util/iovec.rs]: > The latter approach actually causes problems.

Sorry, there is a typo. I want to mean the _former_ approach (which copies `IoVec`s one by one when they are used) causes problems, not the latter approach

.> But I didn't find any error number other than EFAULT.

Yes. If any `IoVec` cannot be accessed, we should directly fail with `EFAULT`.

File: kernel/aster-nix/src/util/iovec.rs (hunk 3 above)
Comment (lrh2000) [reply kernel/aster-nix/src/util/iovec.rs]: Thank you for your opinion. Your point makes sense to me. After more thought, I completely agree that `IoVec` should not allocate the kernel buffer.

But after more thought, I also think that the APIs of `IoVec` probably need to  be better revised in another direction to solve https://github.com/asterinas/asterinas/issues/711. Solving #711 can be outside the scope of this PR, but it has some implications on the API design.

Here, `IoVec` is essentially a slice to user memory  (containing the address to the user buffer and the size of the user buffer), so it can be directly passed to smoltcp (e.g., through the closure in [`Socket::recv`](https://docs.rs/smoltcp/latest/smoltcp/socket/tcp/struct.Socket.html#method.recv)) to detect `EFAULT`s occurred in the middle of the user buffer. If we want to do this, we need flexible APIs for the `IoVec`. For example:
 - we need different APIs for truncatable data (e.g., TCP streams)  and non-truncatable data (e.g., UDP datagrams), perhaps with methods like `read`/`write` and `read_exact`/`write_exact`.
 - we should not require the size of the kernel buffer (e.g., the size of a UDP message) to be the same as the size of the user buffer;
 - we should take `&mut self` and advance the pointer to the user message whenever necessary.

I don't think you want to do that in this PR, do you? If the answer is negative, I think we can also accept the current APIs of `IoVec` as a temporary solution to support `sendmsg` and `recvmsg`.

File: kernel/aster-nix/src/net/socket/ip/stream/mod.rs [hunk 6]
```diff
@@ -589,6 +591,63 @@ impl Socket for StreamSocket {
...
+            }
+
+            let buffer = {
+                let mut buffer = vec![0u8; io_vec.len()];
+                io_vec.read_from_user(&mut buffer)?;
+                buffer
+            };
+
+            // Remote address is ignored
+            let sent_bytes = self.sendto(&buffer, None, flags)?;
```
Comment (lrh2000) [inline kernel/aster-nix/src/net/socket/ip/stream/mod.rs:620]: There should be issues that are similar to `recvmsg`, please also check and fix them.

File: kernel/aster-nix/src/net/socket/ip/datagram/mod.rs [hunk 7]
```diff
@@ -296,6 +297,85 @@ impl Socket for DatagramSocket {
...
+        let mut total_bytes = 0;
+        let mut peer_addr = None;
+        for io_vec in IoVecIter::new(&dst) {
+            if io_vec.is_empty() {
+                continue;
+            }
+
+            let (recv_bytes, addr) = {
+                let mut buffer = vec![0u8; io_vec.len()];
+                let (recv_bytes, addr) = self.recvfrom(&mut buffer, flags)?;
```
Comment (lrh2000) [inline kernel/aster-nix/src/net/socket/ip/datagram/mod.rs:357]: For UDP datagrams, `sendmsg`/`recvmsg` should send/receive only _one_ message. But the current implementation sends/receives multiple messages, which is the behavior of `sendmmsg`/`recvmmsg`, not `sendmsg`/`recvmsg`.

File: kernel/aster-nix/src/net/socket/ip/stream/mod.rs (hunk 6 above)
Comment (lrh2000) [reply kernel/aster-nix/src/net/socket/ip/stream/mod.rs:620]: Not sure whether `sendmsg`/`recvmsg` should be atomic. If it needs to be atomic, we cannot invoke `try_recvfrom`/`try_sendto` multiple times for both TCP streams and UDP datagrams, unfortunately.

Note that `readv`/`writev` must be atomic, according to their [man pages](https://man7.org/linux/man-pages/man2/readv.2.html), as quoted below. But I didn't find similar words for `sendmsg`/`recvmsg`.
> The data transfers performed by readv() and writev() are atomic: [ .. ]

File: kernel/aster-nix/src/util/iovec.rs (hunk 3 above)
Comment (StevenJiang1110) [reply kernel/aster-nix/src/util/iovec.rs]: Thanks for your advice!

> I don't think you want to do that in this PR, do you?

Yes, the problem seems to be out of the scope of this PR. It may require a lot of API changes and I don't want to include them in this PR.

> Here, IoVec is essentially a slice to user memory (containing the address to the user buffer and the size of the user buffer), so it can be directly passed to smoltcp (e.g., through the closure in [Socket::recv](https://docs.rs/smoltcp/latest/smoltcp/socket/tcp/struct.Socket.html#method.recv))

Quite agree. If so, we can directly copy data from user buffer to the socket buffer, instead of creating temporary kernel buffers and do two redundant data copy. But I can't think up of a suitable abstraction for smoltcp to accept user buffer. If you have some idea, we can open a seperate issue to discuss about it.

**Comment by StevenJiang1110** [comment]
> Not sure whether sendmsg/recvmsg should be atomic. If it needs to be atomic, we cannot invoke try_recvfrom/try_sendto multiple times for both TCP streams and UDP datagrams, unfortunately.

Yes, it should be atomic. Referring to [Linux implementation](https://elixir.bootlin.com/linux/v6.0.9/source/net/ipv4/tcp.c#L2679), the recvmsg always be called after locking the socket.

> Note that readv/writev must be atomic, according to their [man pages](https://man7.org/linux/man-pages/man2/readv.2.html), as quoted below.

For the atomicity of readv and writev, I will do it in a following PR.

**Comment by StevenJiang1110** [comment]
I have also encountered an annoying problem when trying sending two messages to the same TCP socket and receiving both messages at once. I have opened an issue #819 for it. The case is added to the last lines of `tcp_err.c` but is currently commented out.

I currently cannot think up of good approach for the problem, so I hope to leave it as future PRs.

**Comment by StevenJiang1110** [comment]
> I have also encountered an annoying problem when trying sending two messages to the same TCP socket and receiving both messages at once. I have opened an issue #819 for it. 

Updated: The problem has been fixed by latest commit. The problem is that the background polling thread is not woken up when `old_instant` is zero. While small packets may be delayed to transmit, the background polling thread is responsible to tranmit the packet.

File: kernel/aster-nix/src/net/socket/ip/stream/mod.rs [hunk 8]
```diff
@@ -84,8 +101,8 @@ impl StreamSocket {
             let pollee = Pollee::new(IoEvents::empty());
             init_stream.init_pollee(&pollee);
             Self {
-                options: RwLock::new(OptionSet::new()),
-                state: RwLock::new(Takeable::new(State::Init(init_stream))),
+                options: RwMutex::new(OptionSet::new()),
+                state: RwMutex::new(Takeable::new(State::Init(init_stream))),
```
Comment (lrh2000) [inline kernel/aster-nix/src/net/socket/ip/stream/mod.rs:105]: While I agree that changing this to mutex might be right (and unavoidable if we want to copy from the user buffer directly to the socket buffer), I don't see the value of changing it in this PR?

File: kernel/aster-nix/src/net/socket/ip/stream/mod.rs [hunk 9]
```diff
@@ -489,12 +498,13 @@ impl Socket for StreamSocket {
...
-            self.try_recvfrom(buf, flags)
+        let recv_bytes = if self.is_nonblocking() {
+            self.try_recvfrom(buf, flags)?
         } else {
-            self.wait_events(IoEvents::IN, || self.try_recvfrom(buf, flags))
-        }
+            self.wait_events(IoEvents::IN, || self.try_recvfrom(buf, flags))?
+        };
+
+        Ok(recv_bytes)
```
Comment (lrh2000) [inline kernel/aster-nix/src/net/socket/ip/stream/mod.rs:507]: Well, I don't understand why this PR removes `poll_ifaces()` and defines `recv_bytes`. Neither change makes sense in the context of this PR.

For the `poll_ifaces()`-related change, I think we can do it in https://github.com/asterinas/asterinas/pull/815. So this PR does not need to touch this method, right?

File: kernel/aster-nix/src/net/socket/ip/stream/mod.rs [hunk 10]
```diff
@@ -651,6 +661,45 @@ impl Socket for StreamSocket {
...
+        let (recv_bytes, _) = self.recvfrom(&mut packet_buffer, flags)?;
+        let copied_bytes = {
+            let packet = &packet_buffer[..recv_bytes];
+            copy_packet_to_user(&dst, packet)
+        };
+
+        // TODO: support receiving control message
+
+        // According to <https://elixir.bootlin.com/linux/v6.0.9/source/net/ipv4/tcp.c#L2645>,
+        // peer address is ignored for connected socket, so we set peer addr as `None``.
```
Comment (lrh2000) [inline kernel/aster-nix/src/net/socket/ip/stream/mod.rs:698]:
```suggestion
        // peer address is ignored for connected socket, so we set peer addr as `None`.
```

File: kernel/aster-nix/src/util/iovec.rs [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+    /// Reads bytes from the user space buffer pointed by
+    /// the `IoVec` to `dst`.
+    /// If successful, the read length will be equal to `dst.len()`.
+    ///
+    /// # Panic
+    ///
+    /// This method will panic if
+    /// 1.`dst.len()` is not the same as `self.len()`;
+    /// 2. `self.is_empty()` is `true`.
+    pub fn read_exactly_from_user(&self, dst: &mut [u8]) -> Result<()> {
```
Comment (lrh2000) [inline kernel/aster-nix/src/util/iovec.rs:68]: I think to be consistent with [`Read::read_exact`](https://doc.rust-lang.org/std/io/trait.Read.html#method.read_exact) provided by the Rust standard library, we should call this method `read_exact_from_user`.

File: regression/apps/network/udp_err.c [hunk 12]
```diff
@@ -198,3 +199,70 @@ FN_TEST(connect)
 	TEST_SUCC(connect(sk_connected, psaddr, addrlen));
 }
 END_TEST()
+
+int set_blocking(int sockfd)
```
Comment (lrh2000) [inline regression/apps/network/udp_err.c:203]: I think we can write:
```c
void set_blocking(int sockfd)
{
	int flags = CHECK(fcntl(sockfd, F_GETFL, 0));
	CHECK(fcntl(sockfd, F_SETFL, flags & (~O_NONBLOCK)));
}
```

File: regression/apps/network/udp_err.c [hunk 13]
```diff
@@ -198,3 +199,70 @@ FN_TEST(connect)
...
+		return -1;
+	}
+
+	return 0;
+}
+
+FN_TEST(sendmsg_and_recvmsg)
+{
+	CHECK(set_blocking(sk_connected));
+	CHECK(set_blocking(sk_bound));
```
Comment (lrh2000) [inline regression/apps/network/udp_err.c:223]: This changes the global state, so I prefer to add a new setup function right before this test:

```c
FN_SETUP(enter_blocking_mode)
{
	set_blocking(sk_connected);
	set_blocking(sk_bound);
}
END_SETUP()
```

File: kernel/aster-nix/src/net/socket/ip/stream/mod.rs (hunk 9 above)
Comment (StevenJiang1110) [reply kernel/aster-nix/src/net/socket/ip/stream/mod.rs:507]: The `poll_ifaces()`-related change was removed from this PR.

File: kernel/aster-nix/src/net/socket/ip/stream/mod.rs (hunk 8 above)
Comment (StevenJiang1110) [reply kernel/aster-nix/src/net/socket/ip/stream/mod.rs:105]: I have removed the change from this PR.

I used to want hold the read lock when `wait_events`, just like what linux does. But since this change is not covered by this PR, the lock change should also not be included.

File: kernel/aster-nix/src/net/socket/ip/stream/mod.rs [hunk 14]
```diff
@@ -63,6 +66,20 @@ enum State {
     Listen(ListenStream),
 }
 
+impl State {
+    fn get_connected(&self) -> Result<&ConnectedStream> {
```
[the remark this answers is not in the document]
Comment (StevenJiang1110) [reply kernel/aster-nix/src/net/socket/ip/stream/mod.rs]: The method has been renamed.

Yes, the method is used only once. I just think extracting the code to a seperate method makes it look more clear.

File: kernel/aster-nix/src/net/socket/ip/datagram/mod.rs [hunk 15]
```diff
@@ -313,6 +317,44 @@ impl Socket for DatagramSocket {
...
+
+        if control_message.is_some() {
+            // TODO: support sending control message
+            warn!("sending control message is not supported");
+        }
+
+        let packet = copy_packet_from_user(&io_vecs);
+
+        let sent_bytes = self.sendto(&packet, addr, flags)?;
+        poll_ifaces();
```
Comment (lrh2000) [inline kernel/aster-nix/src/net/socket/ip/datagram/mod.rs:336]: This line should be removed because `sendto` will poll interfaces.

File: kernel/aster-nix/src/net/socket/unix/stream/socket.rs [hunk 16]
```diff
@@ -273,6 +277,36 @@ impl Socket for UnixStreamSocket {
...
+        } = msg_hdr;
+
+        let packet = copy_packet_from_user(&io_vecs);
+        let sent_len = self.sendto(&packet, addr, flags)?;
+
+        Ok(sent_len)
+    }
+
+    fn recvmsg(&self, dst: Box<[IoVec]>, flags: SendRecvFlags) -> Result<(usize, MessageHeader)> {
+        debug_assert!(flags.is_all_supported());
```
Comment (lrh2000) [inline kernel/aster-nix/src/net/socket/unix/stream/socket.rs:295]: I have found a number of consistency issues between different implementations of `sendmsg/recvmsg`, including but not limited to the following
 - `debug_assert!(flags.is_all_supported());`
 - `warn!("receiving control message is not supported");`
 - not really important, but how you put empty lines between function calls really varies a lot.

For the current implementation, I propose to unify all these implementations into two methods, regardless of whether the socket is TCP/UDP/Unix. How would you like to just make them the default implementations of `Socket`?

In fact, I think we will eventually do the reverse: make `sendto()`/`recvto()` call `sendmsg()`/`recvmsg()` by default, since the functionality of the latter is a superset of the former (and I think Linux already does this, though I have not checked the source). I think it's okay to do it in this PR, how do you feel about it?

File: kernel/aster-nix/src/net/socket/unix/stream/socket.rs [hunk 17]
```diff
@@ -15,26 +15,30 @@ use crate::{
...
     }
 
     pub(super) fn new_listen(listen: Listener) -> Self {
-        Self(RwLock::new(State::Listen(Arc::new(listen))))
+        Self(RwMutex::new(State::Listen(Arc::new(listen))))
     }
 
     pub(super) fn new_connected(connected: Connected) -> Self {
-        Self(RwLock::new(State::Connected(Arc::new(connected))))
+        Self(RwMutex::new(State::Connected(Arc::new(connected))))
```
Comment (lrh2000) [inline kernel/aster-nix/src/net/socket/unix/stream/socket.rs:41]: Well, it seems that the Unix socket implementation is a copy of the early TCP/UDP implementation, and has diverged greatly as TCP/UDP sockets have been heavily refactored.

I'm not familiar with this, so is this change desired and what's the reason for it?

File: kernel/aster-nix/src/net/socket/unix/stream/socket.rs (hunk 16 above)
Comment (StevenJiang1110) [reply kernel/aster-nix/src/net/socket/unix/stream/socket.rs:295]: > In fact, I think we will eventually do the reverse: make sendto()/recvto() call sendmsg()/recvmsg() by default, since the functionality of the latter is a superset of the former (and I think Linux already does this, though I have not checked the source). I think it's okay to do it in this PR, how do you feel about it?

Sure, the latter makes more sense. `sendto()`/`recvto()` should call `sendmsg()`/`recvmsg()` by default. I will refactor the code accordingly.

**Comment by StanPlatinum** [comment]
This PR will help #703.

File: kernel/aster-nix/src/net/socket/unix/stream/socket.rs (hunk 17 above)
Comment (StevenJiang1110) [reply kernel/aster-nix/src/net/socket/unix/stream/socket.rs]: It is undesired and is removed from this PR

**Comment by StevenJiang1110** [comment]
The fix for #819 has been removed from this PR. Adopting the original fix would cause the `send_buf_full` test case in the main branch to fail. 

Considering that the problem is not directly related to this PR, I will investigate the actual cause and address the issue in a future PR.

File: kernel/aster-nix/src/net/socket/util/message_header.rs [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+use super::socket_addr::SocketAddr;
+use crate::{prelude::*, util::IoVec};
+
+/// Message header used for sendmsg/recvmsg
+#[derive(Debug)]
+pub struct MessageHeader {
+    pub(in crate::net) addr: Option<SocketAddr>,
+    pub(in crate::net) io_vecs: Box<[IoVec]>,
+    pub(in crate::net) control_message: Option<ControlMessage>,
+}
```
Comment (lrh2000) [inline kernel/aster-nix/src/net/socket/util/message_header.rs:12]: I think we should have
```rust
#[derive(Debug)]
pub struct MessageHeader {
    pub(in crate::net) addr: Option<SocketAddr>,
    pub(in crate::net) control_message: Option<ControlMessage>,
}
```
and
```rust
fn sendmsg(&self, io_vecs: &[IoVec], msg_hdr: MessageHeader, flags: SendRecvFlags) -> Result<usize>;
fn recvmsg(&self, io_vecs: &mut [IoVec], flags: SendRecvFlags) -> Result<(usize, MessageHeader)>;

fn sendto(&self, io_vecs: &[IoVec], remote: Option<SocketAddr>, flags: SendRecvFlags) -> Result<usize>;
fn recvfrom(&self, io_vecs: &mut [IoVec], flags: SendRecvFlags) -> Result<(usize, Option<SocketAddr>)>;
```

This way,
 - The input and output arguments of `sendmsg` and `recvmsg` are clearer (`io_vecs` should not be included in the return value of `recvmsg`).
 - The extra heap allocation when implementing `recvfrom` and `sendto` (`vec![user_buf].into_boxed_slice()`) can be avoided.
 - `` FIXME: peer_addr should be `None` for recvmsg `` can be addressed. I don't think the Linux implementation writes the remote address in `recvfrom` (maybe you can write a test case to verify this).

File: kernel/aster-nix/src/net/socket/mod.rs [hunk 19]
```diff
@@ -63,18 +63,39 @@ pub trait Socket: FileLike + Send + Sync {
...
     }
 
-    /// Receive a message from a socket
-    fn recvfrom(&self, buf: &mut [u8], flags: SendRecvFlags) -> Result<(usize, SocketAddr)> {
-        return_errno_with_message!(Errno::EOPNOTSUPP, "recvfrom() is not supported");
-    }
-
-    /// Send a message on a socket
+    /// Sends a message on a socket
     fn sendto(
```
Comment (lrh2000) [inline kernel/aster-nix/src/net/socket/mod.rs:67]: Just realized that maybe we should just move the `sendto`/`recvfrom` logic in this file to `syscall/sendto.rs` and `syscall/recvfrom.rs`?

As strong evidence, [the `struct proto` in the Linux kernel](https://elixir.bootlin.com/linux/v6.9.5/source/include/net/sock.h#L1200) has only `recvmsg` and `sendmsg` methods and _no_ `recvfrom` and `sendto` methods, which means that no protocol can override the default behavior of `recvfrom` and `sendto`. So it is safe for us to do the same.

File: kernel/aster-nix/src/util/iovec.rs [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+        assert_eq!(src.len(), self.len);
+        assert!(!self.is_empty());
+
+        write_bytes_to_user(self.base, src)
+    }
+
+    /// Reads bytes from the user space buffer pointed by
+    /// the `IoVec` to `dst`.
+    /// If successful, returns the length of actually read bytes.
+    pub fn read_from_user(&self, dst: &mut [u8]) -> Result<usize> {
```
Comment (lrh2000) [inline kernel/aster-nix/src/util/iovec.rs:99]: Are `read_from_user` and `write_to_user` useful? It seems that they are not used in this PR?

File: kernel/aster-nix/src/net/socket/ip/stream/connected.rs [hunk 21]
```diff
@@ -67,10 +68,11 @@ impl ConnectedStream {
         }
     }
 
-    pub fn try_sendto(&self, buf: &[u8], flags: SendRecvFlags) -> Result<usize> {
+    pub fn try_send(&self, packet: &[u8], flags: SendRecvFlags) -> Result<usize> {
```
Comment (lrh2000) [inline kernel/aster-nix/src/net/socket/ip/stream/connected.rs:71]: My sense is that universally converting the name of the buffer argument in `{try_,}send` to `packet` looks a bit odd, because:
 - We are not necessarily sending a packet. In TCP stream, the user data of multiple system calls can be combined into one packet, and the user data of one system call can be split into multiple packets.
 - So the concept of "packet" does not exist at this point. It will only exist when smoltcp creates outgoing packets from the buffered data.

It _might_ be appropriate to call this argument "packet" in UDP sockets, if emphasizing that they are packets outweighs consistency with TCP sockets. But changing only `send` and skipping `recv` creates another sense of oddity, IMO.

**Review by lrh2000 — COMMENTED** [review]
Thanks for your efforts on this PR. Most of the changes look good to me. Below are the final comments, hopefully.

File: kernel/aster-nix/src/net/socket/mod.rs (hunk 19 above)
Comment (StevenJiang1110) [reply kernel/aster-nix/src/net/socket/mod.rs]: > Just realized that maybe we should just move the sendto/recvfrom logic in this file to syscall/sendto.rs and syscall/recvfrom.rs?

It makes sense to me. The problem is fixed.

> Also, FIXME: peer_addr should be `None` for recvmsg can be addressed. I don't think the Linux implementation writes the remote address in recvfrom for TCP sockets (maybe you can write a test case to verify this).

Yes, I verified that recvfrom does not writes remote address. The FIXME is removed.

File: kernel/aster-nix/src/util/iovec.rs (hunk 20 above)
Comment (StevenJiang1110) [reply kernel/aster-nix/src/util/iovec.rs:100]: The `write_to_user` is used in the function `copy_message_to_user` (originally named `copy_packet_to_user`)

```rust
pub fn copy_message_to_user(io_vecs: &[IoVec], message: &[u8]) -> usize {
    let mut total_bytes = 0;

    for io_vec in io_vecs {
        if io_vec.is_empty() {
            continue;
        }

        let len = io_vec.len().min(message.len() - total_bytes);
        if len == 0 {
            break;
        }

        let src = &message[total_bytes..total_bytes + len];
        match io_vec.write_to_user(src) {
            Ok(len) => total_bytes += len,
            Err(e) => {
                warn!("fails to copy message to user");
                break;
            }
        }
    }

    total_bytes
}
```

`read_from_user` was added as a counterpart function to `write_to_user` but `read_from_user` is never used.

File: kernel/aster-nix/src/syscall/writev.rs [hunk 22]
```diff
@@ -33,19 +20,21 @@ fn do_sys_writev(fd: FileDesc, io_vec_ptr: Vaddr, io_vec_count: usize) -> Result
...
-            let base = io_vec.base;
-            let len = io_vec.len;
-            let mut buffer = vec![0u8; len];
-            read_bytes_from_user(base, &mut buffer)?;
+            let mut buffer = vec![0u8; io_vec.len()];
+            io_vec.read_exact_from_user(&mut buffer)?;
             buffer
         };
+
         let write_len = file.write(&buffer)?;
```
Comment (lrh2000) [inline kernel/aster-nix/src/syscall/writev.rs:38]: In case that the https://github.com/asterinas/asterinas/pull/702#issuecomment-2100167547 comment is forgotten, add a FIXME comment about atomicity.

**Review by lrh2000 — COMMENTED** [review]
Just a few minor style comments.

File: kernel/aster-nix/src/util/iovec.rs [hunk 23]
```diff
@@ -0,0 +1,131 @@
...
+    /// Writes bytes from the `src` buffer
+    /// to the user space buffer pointed by the `IoVec`.
+    ///
+    /// If successful, the written length will be equal to `src.len()`.
+    ///
+    /// # Panics
+    ///
+    /// This method will panic if  
+    /// 1. `src.len()` is not the same as `self.len()`;  
+    /// 2. `self.is_empty()` is `true`.
```
Comment (lrh2000) [inline kernel/aster-nix/src/util/iovec.rs:90]: Trailing spaces should be removed. I'm surprised CI doesn't seem to complain about this.

```suggestion
    /// This method will panic if
    /// 1. `src.len()` is not the same as `self.len()`;
    /// 2. `self.is_empty()` is `true`.
```

**Review by lrh2000 — APPROVED** [review]
LGTM, except for some annoying short read/write problems. These issues are being tracked separately at https://github.com/asterinas/asterinas/issues/711, and this PR is not appropriate to address them.

File: kernel/aster-nix/src/util/iovec.rs [hunk 24]
```diff
@@ -0,0 +1,129 @@
...
+    ///
+    /// If successful, returns the length of actually written bytes.
+    pub fn write_to_user(&self, src: &[u8]) -> Result<usize> {
+        let len = self.len.min(src.len());
+        write_bytes_to_user(self.base, &src[..len])?;
+        Ok(len)
+    }
+}
+
+/// Copies IO vectors from user space
```
[the remark this answers is not in the document]
Comment (StevenJiang1110) [reply kernel/aster-nix/src/util/iovec.rs]: I scan the usage in Rust std, I think that one-sentence doc comments for methods and structs should end with periods. This problem is fixed in this PR.

While it comes for the doc comments for enum variants(or maybe struct fields), the Rust std does not adopt a consistent style. For example, the comments of [std::option::Option](https://doc.rust-lang.org/src/core/option.rs.html#571) ends with periods.
```rust
pub enum Option<T> {
    /// No value.
    None,
    /// Some value of type `T`.
    Some(T),
}
```

While the doc comments of [std::result::Result](https://doc.rust-lang.org/src/core/result.rs.html#502) do not
```rust
pub enum Result<T, E> {
    /// Contains the success value
    Ok(T),
    /// Contains the error value
    Err(E),
}
```

So the doc comments of enum variants and struct fields are left unchanged.

[the remark this answers is not in the document]
Comment (lrh2000) [reply kernel/aster-nix/src/util/iovec.rs]: Thanks for the explanation! :+1:

**Review by tatetian — APPROVED** [review]
LGTM. Thanks for the contribution!
